### PR TITLE
Always update object properties if the property descriptor is applicable

### PR DESCRIPTION
**Pull request by anba**
https://bugs.ecmascript.org/show_bug.cgi?id=3508 contains the background information for this change. And as mentioned in https://bugs.ecmascript.org/show_bug.cgi?id=3508#c8, this change needs to be reviewed very carefully.

### CODE REVIEW

**Comment by anba** [comment]
In addition to the test case in the bugzilla report, here's another variation using SIMD (requires SIMD value semantics!).
```js
function simdNaN(v) {
  var ta = new Float32Array(new Int32Array([0x7fc00000 | v, 0, 0, 0]).buffer);
  return SIMD.Float32x4.load(ta, 0);
}
function nanValue(simd) {
  var ta = new Float32Array(4);
  SIMD.Float32x4.store(ta, 0, simd);
  return new Int32Array(ta.buffer)[0] >>> 0;
}

print(nanValue(simdNaN(0)).toString(16)); // "7fc00000"
print(nanValue(simdNaN(1)).toString(16)); // "7fc00001"
print(Object.is(simdNaN(0), simdNaN(1))); // "true" with SIMD value semantics

var o = {};
o.a = simdNaN(0);
o.a = simdNaN(1);
// "7fc00000" per ES2015/2016, "7fc00001" with this PR applied
print(nanValue(o.a).toString(16));
```

**Comment by littledan** [comment]
Sorry, I haven't reviewed this in detail yet (it's hard to think about!), but another possible future change in this area is to change the NaN canonicalization logic to allow a NaN to switch to any other NaN after operations like `+` or reading from a TypedArray, rather than putting the canonicalization in writing to a TypedArray. This would let SIMD use plain old ECMAScript floats in a list to represent its vector contents, and generally be more explicit and map more to how implementations do things and users think about things. The idea, which IIRC reached consensus at TC39, was to make this change as part of the SIMD.js integration once it reaches Stage 4. I'm not sure if how that change would affect this patch.

**Comment by bterlson** [comment]
@allenwb and @erights it would be good if you could review this closely.

**Comment by erights** [comment]
Unless I misunderstand what this means, it looks wrong. What are the implications of this function returning true? What are the implications of it returning false?

**Comment by anba** [comment]
@erights `ValidateAndApplyPropertyDescriptor` returns `false` when the requested property descriptor update cannot be applied. Otherwise it returns `true`. That means it returns `true` if
  1. The property update does not lead to any observable changes when compared using `SameValue` semantics.
  1. Or the property update can be applied and it leads to an observable change when compared using `SameValue` semantics.

The important bit is the comparison using `SameValue` semantics. Before TypedArrays were introduced into the language, all language types which are observably different returned `false` when compared with `SameValue`. This premise is no longer true in ES2015, because it's possible to observe the difference between two distinct NaN values (NaNs with different bit patterns, e.g. 7ff8000000000000 and fff8000000000000), even though `SameValue` returns `true` when applied on these values.

```js
// Extracts the bit pattern from a number value.
function numberToRawBits(v) {
  var isLittleEndian = new Uint8Array(new Uint16Array([1]).buffer)[0] === 1;
  var reduce = Array.prototype[isLittleEndian ? 'reduceRight' : 'reduce'];
  var uint8 = new Uint8Array(new Float64Array([v]).buffer);
  return reduce.call(uint8, (a, v) => a + (v < 16 ? "0" : "") + v.toString(16), "");
}

// Sample values with different NaNs, tested in Edge, JSC, V8 and SpiderMonkey.
[0/0, NaN, Math.sqrt(-1), Math.hypot(Infinity)].map(numberToRawBits).forEach(v => print(v));
```


### What is the motivation for this change? ###

Simple property assignments like `obj.prop = value` are performed using the ordinary [[Set]] abstract operation, [[Set]] calls the ordinary [[DefineOwnProperty]] operation which in turn calls `OrdinaryDefineOwnProperty` and that one calls `ValidateAndApplyPropertyDescriptor`. As currently specified, this requires implementations to compare the current property value with the new property value using `SameValue` on every property assignment, even for the common case of own writeable properties. And that's probably not great for performance reasons.

```js
var nan1 = /* NaN value */;
var nan2 = /* NaN value, but with a different bit pattern than nan1 */;

var obj = {};
obj.prop = nan1;
obj.prop = nan2;

// ES2015 requirement: `obj.prop` must not be changed to nan2.
assertEq(numberToRawBits(obj.prop), numberToRawBits(nan1));
```


### Impact of the proposed change? ###

`ValidateAndApplyPropertyDescriptor` applies a property update even when the update does not lead to an observable difference when compared using `SameValue` semantics. 


### Alternatives? ###

A possible alternative to this proposal is to remove the capability to observe different NaN bit patterns. But that'd require NaN canonicalization on TypedArray writes which was deemed unacceptable because it hurts performance (source: [es-discuss](https://encrypted.google.com/search?q=site:esdiscuss.org+nan+canonicalization&gbv=1)).

It's also possible to replace step 4 of 9.1.6.3 ValidateAndApplyPropertyDescriptor with:

---
1. If every field in _Desc_ also occurs in _current_ and the value of every field in _Desc_ is the same value as the corresponding field in _current_ when compared using the SameValue algorithm, then
  1. If _O_ is not *undefined* and the [[Value]] field of _Desc_ is present, then
    1. If the [[Writable]] field of _current_ is *true*, then
      1. Set the [[Value]] attribute of the property named _P_ of object _O_ to _Desc_.[[Value]].
  1. Return *true*.

---

But that approach feels a bit hackish to me.

**Comment by bterlson** [comment]
Here's the output from Chakra, D8, and SM:

#### spidermonkey
7ff8000000000000
7ff8000000000000
fff8000000000000
7ff0000000000000

#### chakra
fff8000000000000
fff8000000000000
fff8000000000000
7ff0000000000000

#### d8
7ff8000000000000
7ff8000000000000
fff8000000000000
7ff0000000000000

**Comment by bterlson** [comment]
Note that Chakra only has one NaN representation (7ff... is +Inf I think?)

**Comment by anba** [comment]
> Note that Chakra only has one NaN representation (7ff... is +Inf I think?)

Yup, I've tested Chakra last and got a bit sloppy. Updated the comment accordingly. 

(In Win7+IE11 and Win10+Edge20.10240.16384, `numberToRawBits(NaN)` and `numberToRawBits(-NaN)` can be used to generate different NaNs. Is that still possible in more recent Edge builds?)

**Comment by bterlson** [comment]
> (In Win7+IE11 and Win10+Edge20.10240.16384,  numberToRawBits(NaN)  and  numberToRawBits(-NaN)  can be used to generate different NaNs. Is that still possible in more recent Edge builds?)

(Indeed, this can be reproed in Edge 25 as well. Not in the latest ch.exe but this is likely due to toolchain differences. Nice find!)

**Comment by bterlson** [comment]
I will present on this in Munich to make sure we have consensus for this change, though I think it would be nice to knock it off sooner. I do not have concerns, but I am waiting to get an affirmation from @erights and @littledan at least :)

**Comment by erights** [comment]
@bterlson What program generated the output you show at https://github.com/tc39/ecma262/pull/353#issuecomment-188992534 .

**Comment by bterlson** [comment]
@erights https://github.com/bterlson/eshost-cli (npm install eshost-cli). But you will also need copies of the various runtimes which for the most part you have to build yourself. See more at the readme here: https://github.com/bterlson/eshost.

**Comment by erights** [comment]
@bterlson What I meant is, what does this output signify? What patterns of output are compat with the existing spec? What patterns are compat with the proposed spec change?

Did I miss some context? That message begins with "Here's the output from Chakra, D8, and SM:". Output of what? How were we supposed to interpret these results?

**Comment by bterlson** [comment]
@erights Ahh, sorry I misunderstood. The output is from the program @anba listed above, specifically:

```js
// Extracts the bit pattern from a number value.
function numberToRawBits(v) {
  var isLittleEndian = new Uint8Array(new Uint16Array([1]).buffer)[0] === 1;
  var reduce = Array.prototype[isLittleEndian ? 'reduceRight' : 'reduce'];
  var uint8 = new Uint8Array(new Float64Array([v]).buffer);
  return reduce.call(uint8, (a, v) => a + (v < 16 ? "0" : "") + v.toString(16), "");
}

// Sample values with different NaNs, tested in JSC, V8 and SpiderMonkey.
// Update: Removed Edge and invalid example value 'Math.hypot(Infinity)'
[0/0, NaN, Math.sqrt(-1)].map(numberToRawBits).forEach(v => print(v));
```

**Comment by littledan** [comment]
@jugglinmike has been finding some issues relating to how browsers handle NaN, which seem to be violations of the old spec's "the same value" rule. I'm not sure what we should do with these--I don't know whether they are violations of security properties or simply artifacts of a spec which might be stronger than it needs to be. Let's take these into account when revisiting this part of the spec. Mike, could you provide links to the bugs that you know about?

**Comment by erights** [comment]
@bterlson thanks. I did indeed miss that.

What patterns of output are compat with the existing spec? What patterns of output would be compat with the proposed spec change? In both cases, what might a test look like that could demonstrate lack of conformance to either the current or proposed spec?

**Comment by jugglinmike** [comment]
Sure

- [tc39/test262 #623 - Add tests for NaN handling with Typed
  Arrays](https://github.com/tc39/test262/pull/623)
- [v8 #4996 - NaN value canonicalized
  inconsistently](https://bugs.chromium.org/p/v8/issues/detail?id=4996)
  (observable in Integer Indexed exotic object's [[DefineOwnProperty]])
- [SpiderMonkey #1271646 - NaN value canonicalized
  inconsistently](https://bugzilla.mozilla.org/show_bug.cgi?id=1271646)
  (observable in `%TypedArray%.prototype.slice`)

**Comment by littledan** [comment]
OK, I think this change is acceptable (thanks for the explanation, Mark). If I understand correctly, the change is that we only resort to doing a SameValue check when the field is not Writable, and otherwise we apply the property update. So, if a different NaN value written to a property which is already NaN, then as before this will succeed regardless of whether the property is writable, but after the change, if the property is writable, it will get the new NaN, and if it is not writable, it will remain the old NaN. Is that right? And the SameValue check is more reasonable to do performance-wise in the non-writable case as one doesn't expect lots of writes to happen to non-writable properties.

FWIW in my tests, it seems like V8 is currently a bit inconsistent with following the spec. If you invoke Object.defineProperty, then the SameValue check seems to be done to not redefine to a different NaN value, but an ordinary `=` property write leads to changing the NaN value.

(Aside: Does this mean, then, that we could... remove the "An implementation must always choose the same encoding for each implementation distinguishable NaN value." language from SetValueInBuffer? The language was added to guard against an information leak caused by a possible optimization in SetValueInBuffer. With this change, though, it's rather clear that only writable properties encounter the write, and non-writable properties meet their fate with an early return, regardless of whether SameValue ended up being true. However, this would be a separate, more radical change; the motivation is unrelated.)

@jugglinmike Seems like those tests don't get at the issue described in this post; it might be worth writing another test (though that test may need to be changed if this commit goes in).

**Comment by jugglinmike** [comment]
Sure; here it is https://github.com/tc39/test262/pull/648

**Comment by littledan** [comment]
This change reached consensus at the May 2016 TC39 meeting. @anba could you rebase so we can merge?

**Comment by bterlson** [comment]
I've done the rebase and merged this change!

File: spec.html [hunk 1]
```diff
@@ -6410,20 +6409,22 @@ <h1>ValidateAndApplyPropertyDescriptor (_O_, _P_, _extensible_, _Desc_, _current
               1. If _O_ is not *undefined*, convert the property named _P_ of object _O_ from an accessor property to a data property. Preserve the existing values of the converted property's [[Configurable]] and [[Enumerable]] attributes and set the rest of the property's attributes to their default values.
           1. Else if IsDataDescriptor(_current_) and IsDataDescriptor(_Desc_) are both *true*, then
             1. If the [[Configurable]] field of _current_ is *false*, then
-              1. Return *false*, if the [[Writable]] field of _current_ is *false* and the [[Writable]] field of _Desc_ is *true*.
               1. If the [[Writable]] field of _current_ is *false*, then
+                1. Return *false*, if the [[Writable]] field of _Desc_ is present and the [[Writable]] field of _Desc_ is *true*.
                 1. Return *false*, if the [[Value]] field of _Desc_ is present and SameValue(_Desc_.[[Value]], _current_.[[Value]]) is *false*.
+                1. Return *true*.
```
Comment (ljharb) [inline spec.html]: it seems like this change actually bypasses *applying* the Property Descriptor, in the final step. I think this line should not have been added.

File: spec.html [hunk 2]
```diff
@@ -6410,20 +6409,22 @@ <h1>ValidateAndApplyPropertyDescriptor (_O_, _P_, _extensible_, _Desc_, _current
...
             1. Else the [[Configurable]] field of _current_ is *true*, so any change is acceptable.
           1. Else IsAccessorDescriptor(_current_) and IsAccessorDescriptor(_Desc_) are both *true*,
             1. If the [[Configurable]] field of _current_ is *false*, then
               1. Return *false*, if the [[Set]] field of _Desc_ is present and SameValue(_Desc_.[[Set]], _current_.[[Set]]) is *false*.
               1. Return *false*, if the [[Get]] field of _Desc_ is present and SameValue(_Desc_.[[Get]], _current_.[[Get]]) is *false*.
+              1. Return *true*.
```
Comment (ljharb) [inline spec.html]: Same here.

**Comment by ljharb** [comment]
I've filed https://github.com/tc39/ecma262/pull/1677 to discuss and/or fix.

**Pull request closed without merging**